service/controller/v1/prometheus/scrapeconfig_test.go

```diff
@@ -0,0 +1,503 @@
+
+				TestConfigOneApiserver,
+				TestConfigOneCadvisor,
+				TestConfigOneKubelet,
+				TestConfigOneNodeExporter,
+				TestConfigOneWorkload,
+			},
+		},
+	}
+
+	for index, test := range tests {
+		scrapeConfigs, err := GetScrapeConfigs(test.services, test.certificateDirectory)
+		if err != nil {
+			t.Fatalf("%d: error returned creating scrape configs: %s\n", index, err)
+		}
+
+		if !reflect.DeepEqual(test.expectedScrapeConfigs, scrapeConfigs) {
+			t.Fatalf(
+				"%d: expected scrape configs do not match returned scrape configs.\nexpected: %s\nreturned: %s\n",
+				index,
+				spew.Sdump(test.expectedScrapeConfigs),
+				spew.Sdump(scrapeConfigs),
+			)
+		}
+	}
+}
+
+// Test_Prometheus_GetScrapeConfigs_Deterministic tests that the GetScrapeConfigs function is deterministic,
+// and that scrape configs are returned in alphabetical order by the job name.
+func Test_Prometheus_GetScrapeConfigs_Deterministic(t *testing.T) {
+	services := []v1.Service{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "apiserver",
+				Namespace: "xa5ly",
+				Annotations: map[string]string{
+					ClusterAnnotation: "xa5ly",
+				},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "apiserver",
+				Namespace: "0ba9v",
+				Annotations: map[string]string{
+					ClusterAnnotation: "0ba9v",
+				},
+			},
+		},
+	}
+
+	expectedScrapeConfigs := []config.ScrapeConfig{
+		TestConfigTwoApiserver,
+		TestConfigTwoCadvisor,
+		TestConfigTwoKubelet,
+		TestConfigTwoNodeExporter,
+		TestConfigTwoWorkload,
+
+		TestConfigOneApiserver,
+		TestConfigOneCadvisor,
+		TestConfigOneKubelet,
+		TestConfigOneNodeExporter,
+		TestConfigOneWorkload,
+	}
+
+	for index := 0; index < 50; index++ {
+		scrapeConfigs, err := GetScrapeConfigs(services, "/certs")
+		if err != nil {
+			t.Fatalf("%d: error returned creating scrape configs: %s\n", index, err)
+		}
+
+		if !reflect.DeepEqual(expectedScrapeConfigs, scrapeConfigs) {
+			t.Fatalf(
+				"%d: expected scrape configs do not match returned scrape configs. GetScrapeConfigs not deterministic.\nexpected: %s\nreturned: %s\n",
+				index,
+				spew.Sdump(expectedScrapeConfigs),
+				spew.Sdump(scrapeConfigs),
+			)
+		}
+	}
+}
+
+// Test_Prometheus_YamlMarshal tests that Prometheus marshals yaml correctly.
+func Test_Prometheus_YamlMarshal(t *testing.T) {
+	tests := []struct {
+		service v1.Service
+
+		expectedConfigs string
+	}{
+		{
+			service: v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "apiserver",
+					Namespace: "xa5ly",
+					Annotations: map[string]string{
+						ClusterAnnotation: "xa5ly",
+					},
+				},
+			},
+
+			expectedConfigs: `- job_name: guest-cluster-xa5ly-apiserver
+  scheme: https
+  kubernetes_sd_configs:
+  - api_server: https://apiserver.xa5ly
+    role: endpoints
+    tls_config:
+      ca_file: /certs/xa5ly-ca.pem
+      cert_file: /certs/xa5ly-crt.pem
+      key_file: /certs/xa5ly-key.pem
+      insecure_skip_verify: false
+    namespaces:
+      names: []
```

Comment:
This test here was funky. Some structure changed but nothing major IMO. Would be good to get some pairs of eyes onto this and test this properly in the wild.